message_body: Use grid to arrange elements.

amanagr opened the pull request:
The purpose of this PR is to make messages use grid layout and move away from fixed widths and position for items where possible. There should not be any visual changes in the first commit.

<img width="501" alt="image" src="https://user-images.githubusercontent.com/25124304/200135174-116c2a99-418f-448a-bc17-355d1f83b0d7.png">

discussion: https://chat.zulip.org/#narrow/stream/6-frontend/topic/grid.20fror.20message_body

List of things to check that can go wrong on any change:

- [x] Message alerts like "Copied"
- [x] Edit menu
- [x] Reactions
- [x] Narrow width adjustments
- [x] Action icons when message failed to send
- [x] Next is same sender message everything.
- [x] /me messages

amanagr commented:
Even though I did a lot of testing to ensure silly bugs don't creep in, I am sure there are still some present.

amanagr commented on frontend_tests/node_tests/popovers.js:
```diff
@@ -31,6 +31,7 @@ const message_lists = mock_esm("../../static/js/message_lists", {
 });
```
<img width="445" alt="Screenshot 2022-11-18 at 7 54 09 PM" src="https://user-images.githubusercontent.com/25124304/202726796-99cfa34f-abef-44ad-b47e-37a4c1ca9f0f.png">

amanagr replied:
Notice that username is trimmed too. I am not sure if there is any other solution to that.

amanagr commented on static/styles/message.css:
```diff
@@ -0,0 +1,120 @@
+$first_column_width: 46px;
```
`46px` is we used right now in main too.

timabbott replied:
Let's call this `avatar_column_width`; I think that more clearly expresses its purpose. (`first_column` might make you think of the app's left/middle/right columns).

amanagr replied:
Ok done.

amanagr commented on static/styles/message.css:
```diff
@@ -0,0 +1,120 @@
+.message_row {
+    &.include-sender {
+        .messagebox .messagebox-content {
+            grid-template-rows: 25px repeat(3, auto);
```
`25px` so that we can fit any text here.

timabbott replied:
I don't understand this comment at all.

amanagr replied:
I mean text in different languages can go take a lot of height regardless of line-height or font-size. `25px` seems safe for that.

amanagr commented on static/styles/message.css:
```diff
@@ -0,0 +1,120 @@
+        padding-left: 10px;
+        padding-right: 2px; /* For message focused border */
+        grid-template-rows: repeat(4, auto);
+        grid-template-columns: $first_column_width auto 60px 64px;
```
I had to use these `60` and `64` magic numbers here to fit `message_controls` and time.

amanagr replied:
Okay.

timabbott replied:
Is there a way we can make them a `calc()` function of the sizes used for the message controls regions? That would be a lot more maintainable than comments.

amanagr replied:
124px is where we want to be at for `message_time` + `message_controls` looking at our main branch. So, instead of using these numbers, I set these to `max-content` which works well for mobile widths too. I added padding around message time and message controls to get that desired 124px value on wide widths. Since these elements don't change width with a change in language, we should be fine. Although, time can occupy few pixels more or less width based on what time the message was sent since "10:00" occupies more width than "1:11".

amanagr commented:
There are some `padding` and `margin` used to align content with others, as all elements are not center aligned in every scenario.

alya commented:
This PR fails for /me messages:

<img width="809" alt="Screen Shot 2022-11-22 at 3 06 48 PM" src="https://user-images.githubusercontent.com/2090066/203439266-adb0c4b8-694a-45ff-bc8a-2d7b19278afc.png">

We should make sure the /me editing experience (fixed in #23102) is all good as well.

alya commented:
I didn't come across any other issues in manual testing.

amanagr commented on static/styles/message.css:
```diff
@@ -33,6 +33,25 @@ $first_column_width: 46px;
                 grid-column: 1 / 3;
```
Nice catch!

alya replied:
Hm, it looks like the timestamps and action icons are currently vertically centered rather than being at the top of the message, and I think that looks better than the screenshot above.

amanagr replied:
<img width="664" alt="image" src="https://user-images.githubusercontent.com/25124304/203821257-d12bb719-5c8a-4908-a04f-f59dc80ec15f.png">

Thanks! Fixed!

alya replied:
Thanks! I think the timestamp looks a bit too low in the editing UI for /me messages. Compare these two:

<img width="912" alt="Screen Shot 2022-11-24 at 10 54 02 PM" src="https://user-images.githubusercontent.com/2090066/203919188-42087297-6834-4d3c-9f72-878f2f423b21.png">

Once you're editing, I think it makes sense to align the timestamp at the top, as for other messages.

I checked on CZO, and we actually have the same issue there, so I'm fine with merging this PR and fixing it separately if that works best.

amanagr replied:
Fixed
<img width="647" alt="image" src="https://user-images.githubusercontent.com/25124304/204183494-d9dafb62-5e9f-4775-97ea-526129afa7fc.png">

alya replied:
Thanks! That screenshot looks good to me.

timabbott commented:
We're definitely going to want to test this on chat.zulip.org before trying to merge it, since it's the kind of thing that can break stuff.

timabbott commented:
One thing we'll need to do before we can integrate something like this is profiling at least Chrome's CSS performance for rendering 1K messages before/after this change. I know in the past we've had regressions in the rendering performance of loading 1K messages in Zulip that were caused by subtle CSS details (basically asking the browser to do a lot of work). I have no particularly expectation as to whether this should be faster or slower than the previous code ... but we should test.

amanagr commented:
Here are the results. I think there is a very slight performance improvement.
<img width="617" alt="Screenshot 2022-11-30 at 8 36 24 AM" src="https://user-images.githubusercontent.com/25124304/204698108-16374b38-01e2-445d-a4ad-2ebc70639eca.png">

amanagr commented:
Pushed a fix non-English names not being rendered properly.

alya commented:
This PR fixes a bug I came across in `main`:

![Screen Shot 2022-12-02 at 10 40 58 AM](https://user-images.githubusercontent.com/2090066/205394971-e8203c04-d453-4d91-af96-8e37112cdd6e.png)

@timabbott FYI

amanagr commented on static/js/bundles/app.js:
```diff
@@ -32,6 +32,7 @@ import "../../styles/components.css";
 import "../../styles/app_components.css";
```
![image](https://user-images.githubusercontent.com/25124304/206108694-930c5242-fb4b-41f1-8419-959e2db68ece.png)

amanagr commented on static/js/popover_menus_data.js:
```diff
@@ -1,11 +1,13 @@
 /* This module provides relevant data to render popovers that require multiple args.
```
![Photo 1](https://user-images.githubusercontent.com/25124304/206108790-86ce3511-c743-4185-889e-63bf55b4f511.png)
![Photo 2](https://user-images.githubusercontent.com/25124304/206108844-decaf676-9437-462f-93b6-af0572324f25.png)
![Photo 3](https://user-images.githubusercontent.com/25124304/206108903-a5315c53-c740-4305-9d81-ea1c73c0d2f3.png)

amanagr commented on static/styles/message.css:
```diff
@@ -34,6 +34,31 @@ $distance_of_text_elements_from_message_box_top: 8px;
                 grid-column: 1 / 3;
```
before:
![image](https://user-images.githubusercontent.com/25124304/206109084-412f16cc-cead-42c5-8520-106accc05ee9.png)

after:
![image](https://user-images.githubusercontent.com/25124304/206109143-10c08274-bda5-4aa5-b075-cbf37ff51a11.png)

amanagr commented:
Updated commit message and added images based on discussion https://chat.zulip.org/#narrow/stream/6-frontend/topic/grid.20fror.20message_body. Thanks, @evykassirer, for providing them.

timabbott commented on static/templates/message_body.hbs:
```diff
@@ -1,38 +1,25 @@
+{{/if}}
 
-</div>
+<span class="alert-msg"></span>
```
This lost the `pull-right` class; can you do a screenshot showing a message that contains one of these `alert-msg` elements? I've forgotten what they are.

amanagr replied:
<img width="1721" alt="image" src="https://user-images.githubusercontent.com/25124304/206638760-c386d495-3fdb-4cc4-a7ca-214ccf36bea7.png">

timabbott commented on static/templates/message_body.hbs:
```diff
@@ -48,6 +35,15 @@
+    <span class="copy-paste-text">&nbsp;</span>
+    {{/unless}}
+    {{timestr}}
+</a>
```
Here, we are reordering the timestamp to be after the message body within the message in the HTML. Is this required? I think we intentionally had the timestamp before the message body for copy/paste reasons, but trying to test, it kinda looks like we completely broken being able to copy/paste messages including the timestamps at some point? Or maybe there's some decision I've forgotten; @evykassirer would know given their work in the area.

So maybe there's an unrelated regression here, but I'm curious whether this change in order will make our life difficult for fixing the above bug, and whether there's a way we can avoid it with a few extra lines of CSS.

amanagr replied:
> Here, we are reordering the timestamp to be after the message body within the message in the HTML. Is this required?

No, I reverted them back. I might have been trying something here that I forgot.

timabbott commented on static/styles/message.css:
```diff
@@ -0,0 +1,128 @@
+$first_column_width: 46px;
+$distance_of_text_elements_from_message_box_top: 8px;
+
+.message_row {
```
From a readability standpoint, would it make sense to do `&.include_sender` variations after the base CSS that it overrides?

amanagr replied:
Ok, done!

timabbott commented on static/styles/message.css:
```diff
@@ -0,0 +1,128 @@
+
+                .inline_profile_picture {
+                    flex-shrink: 0;
+                    /* Let user profile take extra height without
```
Do you mean "user profile picture" here? The "user profile" is a modal.

amanagr replied:
Yes. Fixed.

timabbott commented on static/styles/message.css:
```diff
@@ -0,0 +1,128 @@
+            line-height: 1;
+            justify-self: end;
+            position: relative;
+            top: 6px;
```
Is this `6px` the same as the `6px` we have as a `margin-top` for `.inline_profile_picture`? If so, extracting a `include_sender_top_margin` type constant could be very helpful to express that.

amanagr replied:
Okay done!

timabbott commented on static/styles/message.css:
```diff
@@ -0,0 +1,128 @@
+            grid-column-start: 2;
+        }
+    }
+}
```
What do you think about calling this `message_feed.css`, and planning for it to include other CSS that makes up the message feed (stuff like recipient bars) that are not the message body (aka `rendered_markdown.css`)?

I think this name, or perhaps `message_row.css`, are reasonable if we're not going to include some of the CSS for the rest of the message feed.

timabbott replied:
Maybe `message_row.css` is a good scope for a file; there's a lot of things that should probably be moved here from `zulip.css` even with that scope.

amanagr replied:
Yeah, `message_row.css` makes sense to me. Renamed.

timabbott commented on static/styles/zulip.css:
```diff
@@ -933,30 +933,6 @@ td.pointer {
-
-    .message_edit {
-        margin-top: -14px;
-    }
```
It's amazing how many completely random numbers we get to remove here.

timabbott commented on static/js/popover_menus_data.js:
```diff
@@ -89,6 +91,9 @@ export function get_actions_popover_content_context(message_id) {
     const should_display_read_receipts_option =
         page_params.realm_enable_read_receipts && not_spectator;
 
+    const narrow_width = message_viewport.width() < media_breakpoints_num.sm;
```
I think we should call this variable `narrow_message_feed` -- since it's not about the width of the whole window (we have breakpoints where the sidebars hide that impact this calculation, and the math is correct, but the name should communicate clearly what we're checking).

timabbott replied:
Hmm, though thinking about this more, wouldn't a better check be to just look at whether there's in fact a visible add reaction button in the DOM inside the message? If there isn't, we include it in the popover.

That way, we don't have to think about keeping these two implementations in sync.

amanagr replied:
Yeah, makes sense. Added that change.

timabbott commented on static/styles/message.css:
```diff
@@ -60,6 +85,16 @@ $distance_of_text_elements_from_message_box_top: 8px;
+                    & ~ .alert-msg,
+                    & ~ .message_time,
+                    & ~ .message_controls {
+                        grid-row: 1 / 2;
```
This probably deserves a comment for what this `grid-row` thing is doing.

timabbott replied:
Actually, I have a different thought -- I wonder whether we use one of those tools that make nice ASCII diagrams, and put the diagram in a comment at the top of the `message_row` section showing the structure of message rows with all the regions labeled, and then make a separate one for `is_me_message`.

Idunno, maybe that's not helpful, just trying to think about how to make this readable.

amanagr replied:
That's a nice idea. Added the ASCII diagrams.

timabbott commented on static/styles/message.css:
```diff
@@ -34,6 +34,31 @@ $distance_of_text_elements_from_message_box_top: 8px;
+                    & ~ .message_time,
+                    & ~ .message_controls {
+                        grid-row: 1 / 3;
+                    }
```
I think you can squash the /me stuff into the first commit.

amanagr replied:
Done.

timabbott commented:
I think this strategy looks pretty good at a high level. I posted a bunch of comments on details, and a few higher-level issues in the #frontend thread (like "drafts" and "message edit history" being possibly broken by this).

One I noticed about sizing is that in a 960px wide Zulip window (so half a widescreen window), we are reducing the width of a paragraph of content inside the message feed from 400px to 384px with this pull request. I'm not sure I have a clear idea of why that is happening -- are we allocating more space to the `message_controls` or to the timestamp?

amanagr commented on static/styles/message.css:
```diff
@@ -0,0 +1,128 @@
+            grid-row-start: 1;
+            grid-column-start: 2;
+            padding: 2px;
+        }
```
Fixed.

amanagr commented on static/styles/zulip.css:
```diff
@@ -933,30 +933,6 @@ td.pointer {
-
-    .message_content:not(:empty) {
-        margin-top: -18px;
-    }
```
<img width="515" alt="image" src="https://user-images.githubusercontent.com/25124304/206643723-8cbff959-9138-4f29-acdc-890d988283c7.png">
<img width="404" alt="image" src="https://user-images.githubusercontent.com/25124304/206643771-eea61b26-f417-4f88-b9cb-5495b887f26b.png">


Looks like above, I don't we need to do anything here.

timabbott replied:
Great, thanks for checking!

amanagr commented on static/styles/message.css:
```diff
@@ -74,13 +74,24 @@ $distance_of_text_elements_from_message_box_top: 8px;
+                .message_control_button:first-child {
+                    display: none;
+                }
+            }
```
Sounds like a good idea to use 3rd-from-last-child (which is actually 4th). 
<img width="286" alt="image" src="https://user-images.githubusercontent.com/25124304/206647131-e70a9437-3ad8-4ddd-a87f-ef41822f3afb.png">

amanagr commented:
Fixed the drafts and message edit history overlays.
<img width="483" alt="image" src="https://user-images.githubusercontent.com/25124304/206659865-5d6ff467-da6c-44b3-90f2-3496166fcd16.png">
<img width="640" alt="image" src="https://user-images.githubusercontent.com/25124304/206659878-8c1212a5-43a6-443f-9fbb-8d70065f87fd.png">

amanagr commented:
> One I noticed about sizing is that in a 960px wide Zulip window (so half a widescreen window), we are reducing the width of a paragraph of content inside the message feed from 400px to 384px with this pull request. I'm not sure I have a clear idea of why that is happening -- are we allocating more space to the message_controls or to the timestamp?


Before: 
<img width="511" alt="image" src="https://user-images.githubusercontent.com/25124304/206660376-f3f69012-04e8-4395-8ef3-0c2aea628a9f.png">
after:
<img width="506" alt="image" src="https://user-images.githubusercontent.com/25124304/206660737-44ffe0fb-1bcf-44bf-a3be-d8630f20f01b.png">


The new paragraph width can between 483px to 495px based on current time and who is the sender (emoji icon takes more space than move message icon). The reducing in size is due to no overlap between message and the message control icons.

amanagr commented:
Message height comparison:
chat.zulip.org:
<img width="248" alt="image" src="https://user-images.githubusercontent.com/25124304/206662394-35172baf-2a94-4083-9bac-a41752d28336.png">
new:
<img width="253" alt="image" src="https://user-images.githubusercontent.com/25124304/206662448-f1f01756-8b40-4912-bac7-374c30704be8.png">

chat.zulip.org:
<img width="242" alt="image" src="https://user-images.githubusercontent.com/25124304/206664554-c1d6a594-3174-45e2-9ff6-63a445b0192c.png">


new:

<img width="410" alt="image" src="https://user-images.githubusercontent.com/25124304/206664569-03bbdf7f-4932-4c40-9cd8-f9c71e180dd2.png">

timabbott commented:
Overall, this looks pretty good!

I think the width/alignment of the message control still need work. From a design perspective, the star icons should be vertically aligned; and I don't see a reason why we can't have a fixed width for the container for the emoji and move-message icons such that the hover icons are vertically aligned in all messages.

![image](https://user-images.githubusercontent.com/2746074/207460686-983036fa-df84-480f-bcdd-37c8f461d398.png)

Is that something we can do fairly easily? If so, I'd like us to do that adjustment before test-deploying on chat.zulip.org, so that we have a version with no known regressions there.

> The new paragraph width can between 483px to 495px based on current time and who is the sender (emoji icon takes more space than move message icon). The reducing in size is due to no overlap between message and the message control icons.

I'm not totally sure I understand this -- we didn't have visible overlap there in previous versions -- is the point just that we had the margins/padding overlap pretty aggressively in the previous implementation, such that space that one might think of as part of the leftmost icon was overlapping the message body?

amanagr commented:
> I think the width/alignment of the message control still need work. From a design perspective, the star icons should be vertically aligned; and I don't see a reason why we can't have a fixed width for the container for the emoji and move-message icons such that the hover icons are vertically aligned in all messages.

I fixed the width of time column to 60px (based on width it occupies in upstream/main). Added the change in a new commit so that it is easier to review.

amanagr commented:
> I'm not totally sure I understand this -- we didn't have visible overlap there in previous versions -- is the point just that we had the margins/padding overlap pretty aggressively in the previous implementation, such that space that one might think of as part of the leftmost icon was overlapping the message body?

Yes.

timabbott commented:
Test-deploying this on chat.zulip.org.

andersk commented on static/styles/message_row.css:
```diff
@@ -0,0 +1,239 @@
+            padding-top: 2px;
+
+            .message_content {
+                padding: 0;
```
I noticed this as a visual inconsistency between messages with and without senders: `.include-sender` shouldn’t change the _bottom_ padding of the content.

amanagr replied:
Oh, yeah! Fixed to only modify `padding-top:0` for the `include-sender` case.

amanagr commented:
I found that on `upstream/main` alignment of message control icons also varies with the message. So, this could have resulted in me miscalculating a few things.
<img width="427" alt="Screenshot 2022-12-16 at 11 18 46 AM" src="https://user-images.githubusercontent.com/25124304/208031276-2cbc438d-371b-48d4-a68e-1e2f0eb91bc6.png">

So, after using a more robust mechanism, the size of message content is the same as before:
before:
<img width="1476" alt="Screenshot 2022-12-16 at 12 51 24 PM" src="https://user-images.githubusercontent.com/25124304/208045456-546ad8b3-1ee1-4a31-ab9b-b5a38d3f5e27.png">
<img width="1476" alt="Screenshot 2022-12-16 at 12 53 32 PM" src="https://user-images.githubusercontent.com/25124304/208045432-b7f1ba6a-d1db-4684-bb18-b1081e6a8b4b.png">
after:
<img width="1476" alt="Screenshot 2022-12-16 at 12 52 10 PM" src="https://user-images.githubusercontent.com/25124304/208045450-f540465b-3d22-41b7-934e-93496189ca24.png">

<img width="1476" alt="Screenshot 2022-12-16 at 12 50 57 PM" src="https://user-images.githubusercontent.com/25124304/208045465-736ea216-b549-4eb7-b5ef-6cf1ba0b8609.png">

amanagr commented:
Added this change to fix [this bug reported by @timabbott ](https://chat.zulip.org/#narrow/stream/6-frontend/topic/message_row.20css.20redo/near/1485911)

```css
        .include-sender:last-of-type {
            .messagebox .messagebox-content .message_content {
                padding-bottom: 2px;
            }
        }
    }

    .message_row:last-of-type {
        .messagebox .messagebox-content .message_content {
            padding-bottom: 0;
        }
    }
```

timabbott commented on static/styles/message_row.css:
```diff
@@ -0,0 +1,252 @@
+            /* We need to position it from top and not vertically centered since we want it
+               to have the same position from top when user is editing the message. */
+            position: relative;
+            top: $distance_of_non_text_elements_from_message_box_top;
```
So in a single-row message, the message controls look below center by about 1.5px.

![image](https://user-images.githubusercontent.com/2746074/211175536-3716245b-41aa-4d58-95c1-f3200e34b6a1.png)

Changing this to 4.5px rather than 6px seems to fix it, but seeing the above comment, I wonder whether there's a better fix?

amanagr replied:
Yeah, that's the right fix, I forgot to adjust it after adjust this after some padding changes I did as I recall.

timabbott commented:
I tried to verify that the use of space is the same as the pre-PR implementation, and I think it isn't. A few notes:
* A single-line message with no sender indication or recipient bar goes from 28px to 27px in height in this change, according to the browser inspector. The reason I noticed this is that the centering is different -- with a 23px `message_content`, we went from 4px of top padding and 1px of bottom padding to 2px on each side. I think this change affects vertical centering slightly in a way that looks a bit off -- that's why I looked into it.
* As noted above, the message_controls look mis-centered vertically in these single-line messages as well.
* One bug I noticed while reviewing this is that we have this `last_message` CSS class that appears to be applied to every single message if you send several in a row. I haven't looked deeply; this seems to be used just in `zulip.css` and for some code in `lightbox.js`. But I'm not sure it makes any sense that this has a CSS class changing bottom padding; I could imagine that quirk is making alignment issues discussed above inconsistent for testing.

I think the structure of this looks good and expect we should be able to integrate this, but can you do another pass of carefully checking the vertical centering for various classes of messages as either exactly the same as before, or explained with screenshots/etc. in this PR in the chat.zulip.org thread why they're changed. I'd be happy for future changes to this PR to be posted in the form of `[squash]` commits rather than new commits, since I think we're very close.

One detail that does seem to be working correctly is that adjacent paragraphs look the same whether in different messages or the same -- I flag this since it's something that could break while tweaking the above:
![image](https://user-images.githubusercontent.com/2746074/211175917-97618d45-3fba-42c9-a57b-7963eb90f2f6.png)

amanagr commented:
@timabbott pushed a commit to apply the same 4px top and 1px bottom padding for message content as it was before. It seems to be working well. I also removed the last message css since it only had negative effects. So, all single messages are now `28px` in height.

timabbott commented on static/styles/zulip.css:
```diff
@@ -1399,16 +1399,6 @@ td.pointer {
-        padding-bottom: 1px;
-    }
-}
-
```
We should do a follow-up commit to remove the `last_message` computation entirely, I think?

timabbott replied:
I'll do this and merge shortly -- started on it and realized there's a weird lightbox code path to clean up here.

timabbott commented:
I did some more manual testing, and the styling looks great.

OK, I've pushed a couple additional commits related to that `last_message` nonsense, and with those changes, I think this is ready to merge. Thanks for all the work on this @amanagr -- this was a quite complex migration that I think will have dividends for some time to come.

@evykassirer FYI, since I seem to recall you had some exciting work that was waiting on this.
